src/System.Data.OleDb/src/OleDbWrapper.cs

```diff
@@ -0,0 +1,453 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Data.Common;
+using System.Data.ProviderBase;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Data.OleDb {
+    // SafeHandle wrapper around 'DataLinks' object which pools the native OLE DB providers.
+    // expect 1 per app-domain
+    sealed internal class OleDbServicesWrapper : WrappedIUnknown {
+
+        // we expect to store IDataInitialize instance pointer in base.handle
+
+        // since we only have one DataLinks object, caching the delegate here is valid as long we
+        // maintain an AddRef which is also the lifetime of this class instance
+        private UnsafeNativeMethods.IDataInitializeGetDataSource DangerousIDataInitializeGetDataSource;
+
+        // DataLinks (the unknown parameter) is created via Activator.CreateInstance outside of the SafeHandle
+        internal OleDbServicesWrapper(object unknown) : base() {
+            if (null != unknown) {
+                RuntimeHelpers.PrepareConstrainedRegions();
+                try { } finally {
+                    // store the QI result for IID_IDataInitialize
+                    base.handle = Marshal.GetComInterfaceForObject(unknown, typeof(UnsafeNativeMethods.IDataInitialize)); // TODO: this method should be marked with a reliability contract.
```

Comment:
Comment obsolete - we don't use reliability contracts. I guess this will never happen now 😈